rust/stub/src/pe_loader.rs

```diff
@@ -0,0 +1,91 @@
+use goblin::pe::PE;
+use uefi::{
+    prelude::BootServices,
+    table::{
+        boot::{AllocateType, MemoryType},
+        Boot, SystemTable,
+    },
+    Handle, Status,
+};
+
+pub struct Image {
+    pub base: *const u8,
+    pub length: usize,
+    pub entry: extern "efiapi" fn(Handle, SystemTable<Boot>) -> Status,
+}
+
+#[cfg(target_arch = "x86_64")]
+fn flush_instruction_cache(_start: *const u8, _length: usize) {
+    // x86_64 mandates coherent instruction cache
+}
+
+/// Loads and relocates a trusted PE file, returning a structure containing its allocated virtual memory range and entry point.
+pub unsafe fn load(boot_services: &BootServices, file_data: &[u8]) -> uefi::Result<Image> {
```

Comment:
If this function is marked `unsafe` it should have a `/// # Safety` section. See [here](https://rust-lang.github.io/api-guidelines/documentation.html#function-docs-include-error-panic-and-safety-considerations-c-failure).

That being said, I don't think this function should be marked `unsafe` to begin with, because it should always be safe to call. Is there some situations where it's not?

Reply:
The function was pretty wildly memory unsafe due to almost no checks being performed. I added some easy checks, moved the unsafety to starting the image and documented it there.